lib/CL/devices/level0/level0-compilation.cc

```diff
@@ -0,0 +1,674 @@
+
+void Level0CompilationJobScheduler::cancelAllJobsFor(Level0Program *Program) {
+  JobQueue->cancelAllJobsFor(Program);
+}
+
+void Level0CompilationJobScheduler::addCompilationJob(
+    Level0CompilationJobSPtr Job) {
+  JobQueue->pushWork(Job);
+}
+
+Level0CompilationJobScheduler::~Level0CompilationJobScheduler() {
+  JobQueue->clearAndExit();
+  CompilerThreads.clear();
+}
+
+// ***********************************************************************
+// ***********************************************************************
+
+/*
+bool Level0CompilationJobScheduler::createAndWaitForO0Builds(Level0ProgramSPtr
+Program, std::string &BuildLog, bool DeviceSupports64bitBuffers) {
+  Level0ProgramBuildUPtr O0SmallOfsBuild(new Level0ProgramBuild(
+                                                 false, false, false,
+                                                 Program.get()));
+  Level0CompilationJobSPtr O0SmallOfsBuildJob(
+              new Level0CompilationJob(true,
+                    Program, std::move(O0SmallOfsBuild)));
+
+  addCompilationJob(O0SmallOfsBuildJob);
+  O0SmallOfsBuildJob->waitForFinish();
+
+  if (!O0SmallOfsBuildJob->isSuccessful()) {
+    BuildLog.append(Program->getBuildLog());
+    return false;
+  }
+
+  // TODO submit both & wait for both
+  if (DeviceSupports64bitBuffers) {
+      Level0ProgramBuildUPtr O0LargeOfsBuild(new Level0ProgramBuild(
+                                                     false, true, false,
+                                                     Program.get()));
+      Level0CompilationJobSPtr O0LargeOfsBuildJob(
+                  new Level0CompilationJob(true,
+                        Program, std::move(O0LargeOfsBuild)));
+
+      addCompilationJob(O0LargeOfsBuildJob);
+
+      O0LargeOfsBuildJob->waitForFinish();
+      if (!O0LargeOfsBuildJob->isSuccessful()) {
+        BuildLog.append(Program->getBuildLog());
+        return false;
+      }
+  }
+
+  return true;
+}
+
+void Level0CompilationJobScheduler::createO2Builds(Level0ProgramSPtr Program,
+                                                   bool
+DeviceSupports64bitBuffers) { Level0ProgramBuildUPtr O2SmallOfsBuild(new
+Level0ProgramBuild( true, false, false, Program.get()));
+    Level0CompilationJobSPtr O2SmallOfsBuildJob(
+                new Level0CompilationJob(true,
+                      Program, std::move(O2SmallOfsBuild)));
+
+    addCompilationJob(O2SmallOfsBuildJob);
+
+    if (DeviceSupports64bitBuffers) {
+        Level0ProgramBuildUPtr O2LargeOfsBuild(new Level0ProgramBuild(
+                                                       true, true, false,
+                                                       Program.get()));
+        Level0CompilationJobSPtr O2LargeOfsBuildJob(
+                    new Level0CompilationJob(true,
+                          Program, std::move(O2LargeOfsBuild)));
+
+        addCompilationJob(O2LargeOfsBuildJob);
+    }
+}
+*/
+
+bool Level0CompilationJobScheduler::createAndWaitForExactBuilds(
```

Comment:
This requires a method comment since non-trivial.

Reply:
done